### ISSUE TITLE
Flaky Tests Hub

### ISSUE THREAD

**Issue by doc-han** [OpenFn/kit#844]
## Description
This is a space dedicated to tracking known flaky tests and systematically resolving them.

<p align="center">
  <img src="https://gifdb.com/images/high/this-family-guy-fixing-broken-window-shade-et7pgddr5qwtnp7i.webp" width="300" />
</p>

> [!NOTE]
> 1. Pick one, create issue if it doesn't exist and reference here via **related issue**
> 2. After your issue's related PR has been merged and related issue is closed on a good note. click the checkmark!

## Flaky Tests
- [ ] **"synchronously create a placeholder before generating the docs"**
**Reason:** Promise functions start resolving as soon as they've been called. `await` doesn't tell a promise to start resolving but rather blocks execution only in the scope of its async function.
https://github.com/OpenFn/kit/blob/7554a34c009eb722c9e38ead837096e3e249db31/packages/cli/test/docgen/handler.test.ts#L137-L157
**Related Issue:** no issue created
---
- [ ] **"allow a job to complete after receiving a sigterm"**
**Reason:** `workerProcess.kill('SIGTERM')` doesn't kill a process exactly after it has been called. It's async and may take some time to actually kill the process. Hence at times, the job gets processed before, other times not depending on computer resource available. 
https://github.com/OpenFn/kit/blob/535da76adb5ad8b64f095a05804816cdb2da5d84/integration-tests/worker/test/server.test.ts#L98-L145
**Related Issue:** no issue created

**Comment by josephjclark** [comment]
That first one is fine @doc-han - the await at the end is just to close the test out nicely. It's not really significant

**Comment by josephjclark** [comment]
We can skip that sigterm one, it's fine:)

**Comment by josephjclark** [comment]
@doc-han I'm really not sure what's up with the docgen one. It'll occasionally fail locally too, but I can't reliably repro and have yet to see the exact problem. I've spent too much time on this and I've skipped it for now.

There's clearly a timing issue go on but I can't work out where. Yes, the promise starts executing immediately. The docgen function will SYNCHRONOUSLY create a placeholder file, then asynchronously start running doc gen, and synchronously return that promise. It's like the synchronous file write hasn't actually commited to disk by the time we read the file back. Which is weird because fundamentally I expect `writeFileSync('file.txt', 'x'); const f = readFileSync('file.txt', 'utf8')` to reliably work. And as far as I'm concerned, that's what's going on here.

Maybe `mock-fs` is causing some kind of interference.

**Comment by doc-han** [comment]
> @doc-han I'm really not sure what's up with the docgen one. It'll occasionally fail locally too, but I can't reliably repro and have yet to see the exact problem. I've spent too much time on this and I've skipped it for now.

Yeah, I think skipping is the best option now. The only solution is to change how `docsHandler` works. Instead doing two things at the same time, it could be broken into two. 1. handles placeholder generation 2. handles the actual docgen.

**Comment by josephjclark** [comment]
I think mock-fs on node 22 is the real problem. I think the mock is slow, or something, and it's possible to try and read the file before the write has finished in the background, and that's what's triggering these errors.

I'm certainly noticing that tests which use mock-fs are running very slowly.

I also can't seem to reproduce the uncaught error thing by throwing in strategic places. I really don't understand why the error isn't caught.

**Comment by josephjclark** [comment]
```
packages/logger test:   logger › JSON timestamps are bigints representing sensible times
packages/logger test:   test/logger.test.ts:236
packages/logger test:    235:   const endDate = timestampToDate(endTime);                            
packages/logger test:    236:                                                                        
packages/logger test:    237:   // These dates, captured at different resolutions and converte to ms,
packages/logger test:   Value is not `true`:
packages/logger test:   false
packages/logger test:   › file://test/logger.test.ts:236:11
```

**Comment by josephjclark** [comment]
```
packages/engine-multi test:   ✘ [fail]: integration › redact log line state if it exceeds the payload limit Test finished without running any assertions
```

Dang this one is brand new. I wonder why it's flaky?

**Comment by josephjclark** [comment]
This is failing consistently on CI right now :( 4 in row:
```
 ✔ worker › pool › destroy should handle un-initialised workers
packages/engine-multi test:   ✘ Timed out while running tests
packages/engine-multi test:   14 tests were pending in test/worker/pool.test.ts
packages/engine-multi test:   ◌ worker › pool › destroy should close all child processes
packages/engine-multi test:   ◌ worker › pool › destroy gracefully
packages/engine-multi test:   ◌ worker › pool › destroy immediately
packages/engine-multi test:   ◌ worker › pool › throw on exec if destroyed
packages/engine-multi test:   ◌ worker › pool › listen to an event
packages/engine-multi test:   ◌ worker › pool › listen to an event in two successive tasks
packages/engine-multi test:   ◌ worker › pool › listen to an event in two successive tasks after a queue
packages/engine-multi test:   ◌ worker › pool › throw if task times out
packages/engine-multi test:   ◌ worker › pool › throw if a deferred task times out
packages/engine-multi test:   ◌ worker › pool › after timeout, destroy the worker and reset the pool
packages/engine-multi test:   ◌ worker › pool › events should disconnect between executions
packages/engine-multi test:   ◌ worker › pool › returnToPool: add to the start of a full pool
packages/engine-multi test:   ◌ worker › pool › returnToPool: add to the end of an empty pool
packages/engine-multi test:   ◌ worker › pool › returnToPool: add to the middle of a empty pool
```
